### PR TITLE
[Development] Show errors in review

### PR DESCRIPTION
**Pull request by Mottie**
## Description

This is a global form change that shows the user any form errors on the review page after attempting to submit the form.

These changes will not be shown in production until it has been review throughly.

Original ticket: https://github.com/department-of-veterans-affairs/va.gov-team/issues/2138
Original PR: https://github.com/department-of-veterans-affairs/vets-website/pull/11564 (split into smaller PRs)

**WIP**: Things that still need to be done, or fixed (so far)

* <details><summary>Some dynamic required fields.</summary>
  I think this problem is mostly due to issues with the `schema`, but I haven't searched for, or attempted to fix them all. 

  For example, if you test form 526 (see instructions above). Use the "Please select at least one type of supporting evidence" link. The accordion & section will open in edit mode. Select any (or all) of the 3 checkboxes. The error in the alert will disappear, with no new errors, and the sections that were checked will remain visible with a warning icon.</details>

* <details><summary>Missing some dynamic error updates</summary>
  The list of errors do not always update dynamically. This may or may not be due to the schema; I could get this error to show if I made the code search for and include all undefined arrays, but then non-required array entries would be added to the error list.
    
  For example:
  
    * In form 526, use the "New disabilities" error link.
    * After choosing "Yes" and using "Update page", a "New condition" section will appear; but it will not be included in the error list. 
    * Now use the "New condition" edit button and add an entry like "GERD"
    * Use "Update" to save the entry.
    * Now scroll down to the error list. No mention of the condition is listed  _until_ you press the "Submit application" button.
    * You will now see a "First new disabilities cause" error link appear.
  </details>
* <details><summary>Show error messages in edit mode</summary>

  Normally, error messages are not visible until the form element has been focused and blurred, or if the user tries to update or submit the page while the form element is visible. @jenstrickland recommended to make the form errors visible when the page is opened in edit mode. I haven't been able to programmatically make the form errors visible. I have tried:

  * Initially called `props.formContext.onError()` which forced _all_ form errors to show up; now I can't remember where I got the `formContext` that included that function. Either way, showing _all_ errors isn't ideal. 
  * Calling `onBlur` on the element itself - no change; but I think we would first need access to the internal `touched` value.
  * Calling `props.formContext.onBlur(elementID)` - function not accessible from the `RoutedSavableReviewPage` (grandparent component of the SubmitButtons component where the error links code lives)
  </details>

* <details><summary>Add e2e unit tests</summary>
  Once these changes are nearly finalized, e2e tests will be added.</details>

## How to test

* You'll need to have the local server & vets-api running
* <details><summary>For the HLR page:</summary>

  
  * No need to log in - this page is still a WIP
  * Go to http://localhost:3001/decision-reviews/higher-level-review/request-higher-level-review-form-20-0996/review-and-submit
  * Check the "I have read and accept..." checkbox
  * Press "Submit application" button
  * You should now be focused on the error alert box

    <details><summary>Screenshot</summary>

    
    Screen Shot 2020-03-02 at 9 45 33 AM
    </details>
  * The "Schedule times" link won't work because I need to fix the schema's dynamic requirement method. I didn't change it so I could demo examples of broken links.
 </details>

* <details><summary>For form 526EZ:</summary>

  
  * Log in to user 228
  * Go straight to http://localhost:3001/disability/file-disability-claim-form-21-526ez/review-and-submit
  * Click "Continue" on the ITF page
  * Check the "I have read and accept..." checkbox
  * Press "Submit application" button
  * You should now be focused on the error alert box

    <details><summary>Screenshot</summary>

    
    Screen Shot 2020-03-02 at 9 54 40 AM
    </details>
  </details>

* Other forms haven't been tested (yet)

## Testing done

- Local unit tests
- Tests for new/modified components

## Screenshots

The list of interactive errors links in the alert box

* <details><summary>HLR</summary>

  
  Screen Shot 2020-03-02 at 10 13 52 AM

  Note: the "Scheduled times" link will not focus on any form elements, at least until the "Informal conference choice" is made. This was left in place to show that some form errors are due to the structure of the `uiSchema` ; a simple change to the `uiSchema` will prevent this "extra" link from showing up.
  </details>

* <details><summary>Form 526</summary>

  
  Screen Shot 2020-03-02 at 10 14 06 AM
  </details>

## Acceptance criteria
- [ ] Form error links only show up in local, staging & dev environments
- [ ] Error links scroll to the associated page form & focus on the label/legend immediately above the form element (input, radio, etc)
- [ ] After updating, the item is removed from the error list
- [ ] After all items have been handled, the error list message will change to inform the user to try submitting again

## Definition of done
- [x] Events are logged appropriately; error messages are already logged to Sentry.
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by ndsprinkle — APPROVED** [review]

**Comment by Mottie** [comment]
I know this is a big PR, but over 500 lines are unit tests and about another 100 are just comments (docs)

**Comment by meganhkelley** [comment]
VSP and BAM1 are discussing this in Slack: https://dsva.slack.com/archives/C5AGLBNRK/p1585142292008900

Please hold off on this PR until further direction is provided. Thank you!

**Comment by Mottie** [comment]
Updated to only show the form errors for form 526 in a non-production environment.

**Comment by Mottie** [comment]
The content in those error boxes isn't ideal either

**Comment by meganhkelley** [comment]
Per a meeting between BAM 1 and VSP that happened in late March/early April 2020, VSP and design system council recommended that this change be tested with Veterans before it is released or added to the forms library and/or design system.

**Comment by Mottie** [comment]
This code has been re-implemented based on the latest master.

**Comment by ncksllvn** [comment]
@Mottie are the items listed under "WIP: Things that still need to be done, or fixed (so far)" still relevant, or are they done now?

**Comment by Mottie** [comment]
Yes, those issues are still relevant.
- The first two are related, I think, to the use of two separate validation methods. The active validation is done by `react-json-schema` whereas pre-submit another validation is performed using a third party library ([`jsonschema`](https://github.com/tdegrunt/jsonschema)).
- Error messages in edit mode are dependent on passing in a `formContext`, which I'm not sure how to handle (or remember from this almost year-old code)
- End-to-end tests - not done until there's a plan for this to go to production.

**Comment by ncksllvn** [comment]
Okay, cool. Just confirming - this PR is ready for review and merge, correct? Those listed items are just potential areas of improvement to iterate on?

**Comment by Mottie** [comment]
Yes, it's ready.  It'll _only_ be available in staging (for now) to make it easier to evaluate these changes.

**Comment by Mottie** [comment]
Some of the code from this PR has been split out to this smaller one (~700 lines of code with ~400 of that being test data and ~60 lines of comments) - https://github.com/department-of-veterans-affairs/vets-website/pull/14465 - once merged, I will update this PR with the difference

**Review by ndsprinkle — APPROVED** [review]

**Comment by Mottie** [comment]
This PR has been superseded by https://github.com/department-of-veterans-affairs/vets-website/pull/16431

**Pull request closed without merging**
